src/main/scala/rover/rdo/client/RdObject.scala

```diff
@@ -26,42 +28,33 @@ trait RdObject {
   * @param one Some RDO
   * @param other Some other RDO
   */
-class CommonAncestor[RDO <: RdObject](private val one: RDO, private val other: RDO) extends RdObject {
+class CommonAncestor[A](private val one: RdObject[A], private val other: RdObject[A]) extends RdObject[A](null) { // todo: fixme with a deferred state
 
 	// determine it once and defer all RdObject methods to it
-	private val commonAncestor: RDO = {
+	def commonAncestor: RdObject[A] = {
 		// determine here... & probably cache or is that not needed in scala? :S
-		// FIXME: proper determination (need to have whole range of intermediate
-		// versions available
-		one
-	}
-
-	override def currentVersion: Long = {
-		commonAncestor.currentVersion
-	}
-
-	override def stableVersion: Long = {
-		commonAncestor.stableVersion
-	}
-
-	// FIXME: determine what to do with this, fix return value/type
-    def hasDiverged: Long =  {
-		// FIXME: non-logical result
-	    if (this.currentVersion != other.currentVersion){
-			commonAncestor.currentVersion
-		}
-		else{
-			println("Non-divergent objects")
-			this.currentVersion
+		// FIXME: currently we return just the atomic state..what about the outstanding operations?
+		for (i <- one.state.log.asList.reverse) {
+			for (j <- other.state.log.asList.reverse) {
```

Comment:
Double loop seems sub-optimal, is there not a library function we could use here? Maybe some existing binary search function?

Reply:
We will definitely optimize here. But first we need to have a working implementation :)